src/OrchardCore/OrchardCore.Abstractions/Json/Dynamic/JsonDynamicValue.cs

```diff
@@ -1,35 +1,297 @@
+        => value?.JsonValue?.GetValue<ulong>()
+        ?? throw new InvalidCastException($"Cannot convert {value} to UInt64");
+
+    public static explicit operator ulong?(JsonDynamicValue value)
+        => value?.JsonValue?.GetValue<ulong?>();
+
+    public static explicit operator byte[]?(JsonDynamicValue value)
     {
-        get
+        if (value?.JsonValue.GetObjectValue() is string str)
         {
-            if (!_hasValue)
+            return Convert.FromBase64String(str);
+        }
+
+        throw new InvalidCastException($"Cannot convert {value} to Byte array");
+    }
+
+    public static explicit operator TimeSpan(JsonDynamicValue value)
+    {
+        if (value?.JsonValue?.GetObjectValue() is string str)
+        {
+            return TimeSpan.Parse(str, CultureInfo.InvariantCulture);
+        }
+
+        throw new InvalidCastException($"Cannot convert {value} to TimeSpan");
+    }
+
+    public static explicit operator TimeSpan?(JsonDynamicValue value)
+    {
+        var str = value?.JsonValue?.GetObjectValue() as string;
+
+        return str is not null
+        ? TimeSpan.Parse(str, CultureInfo.InvariantCulture)
+        : null;
+    }
+
+    public static explicit operator Uri?(JsonDynamicValue value)
+        => new Uri(value?.JsonValue?.GetValue<string>() ?? string.Empty);
+
+    private sealed class JsonDynamicMetaObject : DynamicMetaObject
+    {
+        private static readonly Dictionary<Type, MethodInfo> _cachedReflectionInfo = typeof(JsonDynamicValue)
+            .GetMethods(BindingFlags.Public | BindingFlags.Instance)
+            .Where(method => method.Name == "op_Explicit")
+            .ToDictionary(method => method.ReturnType);
+
+        public JsonDynamicMetaObject(Expression expression, JsonDynamicValue value)
+            : base(expression, BindingRestrictions.Empty, value)
+        {
+        }
+
+
+        // The 'BindConvert()' method is automatically invoked to handle type conversion when casting
+        // dynamic types to static types in C#. 
+        //
+        // For example, when extracting a 'DateTime' value from a dynamically typed
+        // content item's field:
+        // 
+        // dynamic contentItem = [...]; // Assume contentItem is initialized properly
+        //
+        // 'BindConvert()' is called implicitly to convert contentItem.Content.MyPart.MyField.Value 
+        // to 'DateTime' with similar behavior to the following:
+        // var dateTimeValue = (DateTime)contentItem.Content.MyPart.MyField.Value;
+        public override DynamicMetaObject BindConvert(ConvertBinder binder)
+        {
+            var targetType = binder.Type;
+
+            if (_cachedReflectionInfo.TryGetValue(targetType, out var castMethod))
             {
-                _value = JsonValue?.GetObjectValue();
-                _hasValue = true;
+                var convertExpression = Expression.Convert(Expression.Convert(Expression, typeof(JsonDynamicValue)), targetType, castMethod);
```

Comment:
Shouldn't this expression be cached instead since it varies only by type?

If so you can move the cache initialization code in a static constructor or it might involve unreadable code.

Reply:
@gvkries are you able to address the last comment so we can merge this?

Reply:
These expressions are wrapping another expression from the instance property `Expression`. We cannot cache them here.